internal/status-tracker/src/status_tracker.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::{fmt::Debug, iter::Iterator};
+
+use crate::LogItem;
+
+/// A `StatusTracker` is used in the validation logic of c2pa-rs and
+/// related crates to control error-handling behavior and optionally
+/// aggregate log messages as they are generated.
+#[derive(Debug, Default)]
+pub struct StatusTracker {
+    error_behavior: ErrorBehavior,
+    logged_items: Vec<LogItem>,
+    ingredient_uris: Vec<String>,
+}
+
+impl StatusTracker {
+    /// Returns a [`StatusTracker`] with the specified [`ErrorBehavior`].
+    pub fn with_error_behavior(error_behavior: ErrorBehavior) -> Self {
+        Self {
+            error_behavior,
+            logged_items: vec![],
+            ingredient_uris: vec![],
+        }
+    }
+
+    /// Returns the current list of validation log items.
+    pub fn logged_items(&self) -> &[LogItem] {
+        &self.logged_items
+    }
+
+    /// Returns a list of validation log items that can be mutated if needed.
+    pub fn logged_items_mut(&mut self) -> &mut [LogItem] {
+        &mut self.logged_items
+    }
+
+    /// Appends the contents of another [`StatusTracker`] to this list of
+    /// validation log items.
+    pub fn append(&mut self, other: &StatusTracker) {
+        for log_item in other.logged_items() {
+            self.add_non_error(log_item.clone());
+        }
+    }
+
+    /// Adds a non-error [`LogItem`] to this status tracker.
+    ///
+    /// Primarily intended for use by [`LogItem::success()`]
+    /// or [`LogItem::informational()`].
+    pub fn add_non_error(&mut self, mut log_item: LogItem) {
```

Comment:
Given that LogItem now contains a kind field, there's no need for separate add_non_error and add_error. We should have an add() method that does the right thing. If we aren't ready to update all the calling code, the add_non_error and add_error should simply call add().

Reply:
This is carried over from the existing APIs. The primary difference is that `add_error` _can_ trigger an `Err` result if the status tracker is configured to stop on first error and in that case, we need to have an `Error` type ready to use.

The `add_non_error` case has no return value, so we don't need to pass in an `Error`.